File: Java/Napakalaki/src/napakalaki/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import GUI.Dice;
import java.util.ArrayList;

/**
 *
 * @author juane
 */
public class Player {
    static final int MAXLEVEL=10;
    private  String name;
    private int level;
    private boolean dead=true;
    private boolean canISteal=true;
    protected Player enemy;
    protected ArrayList<Treasure> hiddenTreasures;
    protected ArrayList<Treasure> visibleTreasures;
    private BadConsequence pendingBadConsequence;
    
    public Player(String name){
        this.name= name;
        level=1;
        enemy= null;
        visibleTreasures= new ArrayList();
        hiddenTreasures= new ArrayList();         
    }
    
    public Player(Player p){
        name= p.name;
        level=p.level;
        dead=p.dead;
        canISteal=p.canISteal;
        enemy= p.enemy;
        hiddenTreasures= p.hiddenTreasures;
        visibleTreasures= p.visibleTreasures;
        pendingBadConsequence= p.pendingBadConsequence;
    }
    
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        Dice dice=Dice.getInstance();
        return dice.nextNumber()==6;
    }
    
    protected Player getEnemy(){
        return enemy;
    }
    
    public String getName(){
        return name;
    }
    
    private void bringToLife(){
        dead=false;
    }
    
    public int getCombatLevel(){
        int i;
        int level_combat=level;
        for(i=0; i<visibleTreasures.size(); i++)
            level_combat+=visibleTreasures.get(i).getBonus();
        
        return level_combat;
    }
    
    private void incrementLevels(int i){
        level+=i;
        if(level >= MAXLEVEL)
            level = 10;
    }
    
    private void decrementLevels(int i){
        level-=i;
        if(level <= 0)
            level = 1;
    }
    
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence = b;
    }
    
    private void applyPrize(Monster m){
        int nLevels= m.getLevelsGained();
        int n_treasures= m.getTreasuresGained();
        
        incrementLevels(nLevels);
        
        if(n_treasures>0){
            CardDealer dealer=CardDealer.getInstance();
            Treasure tr_aux;
            for(int i=0; i<n_treasures; i++){
                tr_aux= dealer.nextTreasure();
                hiddenTreasures.add(tr_aux);
            }
        }
    }
    
    private void applyBadConsequence(Monster m){
        BadConsequence badC= m.getBadConsequence();
        decrementLevels(badC.getLevels());
        
        BadConsequence pendingBad= badC.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
        setPendingBadConsequence(pendingBad);
    }
    
    private Boolean canMakeTreasureVisible(Treasure t){
        Boolean puede_make=true;
        int n_tr_oh=0;  //Contamos el nº de tesoros de una mano
        for(Treasure tes :visibleTreasures ){
            
            if(tes.getType()==t.getType() && t.getType() != TreasureKind.ONEHAND)
                return false;
            
            if(tes.getType()==TreasureKind.ONEHAND){
                n_tr_oh++;
                
                if(t.getType()==TreasureKind.BOTHHANDS)
                    return false;
            }
            
            if(tes.getType()==TreasureKind.BOTHHANDS && t.getType()==TreasureKind.ONEHAND)
                return false;
                
        }
        if(t.getType()==TreasureKind.ONEHAND && n_tr_oh>=2)
                return false;
        
        return puede_make;
    }
    
    private int howManyVisibleTreasures(TreasureKind t){
        int n_trs=0;
        for(int i=0; i<visibleTreasures.size(); i++)
            if(visibleTreasures.get(i).getType()==t)
                n_trs++;
        return n_trs;
    }
    
    private void dielfNoTreasures(){
        if(visibleTreasures.isEmpty() && hiddenTreasures.isEmpty())
            dead = true;
    }
    
    public boolean isDead(){
        return dead;
    }
    
    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }
    
    public ArrayList<Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
    
    public BadConsequence getPendingBadConsequence(){
        return pendingBadConsequence;
    }
    
    public CombatResult combat(Monster m){
        CombatResult cr;
        Dice dice;
        int myLevel = getCombatLevel();
        int monsterLevel = getOponentLevel(m);
        if(!canISteal()){
            dice = Dice.getInstance();
            int number = dice.nextNumber();
            
            if(number < 3){
                int enemyLevel = enemy.getCombatLevel();
                monsterLevel += enemyLevel;
            }
        }
        if(myLevel > monsterLevel){
            applyPrize(m);
            cr=(level >= MAXLEVEL)?CombatResult.WINGAME:CombatResult.WIN;
        }
        else{
            cr=shouldConvert()?CombatResult.LOSEANDCONVERT:CombatResult.LOSE;
            applyBadConsequence(m); 
        }
        return cr;
    }
    
    public void makeTreasureVisible(Treasure t){
        Boolean canI=canMakeTreasureVisible(t);
        if(canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t);
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractVisibleTreasure(t);
        dielfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t);
        if(pendingBadConsequence != null && (!pendingBadConsequence.isEmpty()))
            pendingBadConsequence.substractHiddenTreasure(t);
        dielfNoTreasures();
    }
    
    public boolean validState(){
        return (hiddenTreasures.size() < 5) && (pendingBadConsequence == null || pendingBadConsequence.isEmpty());
    }
    
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        bringToLife();
        Treasure t = dealer.nextTreasure();
        hiddenTreasures.add(t);
        int number = dice.nextNumber();
        
        if(number > 1){
            t = dealer.nextTreasure();
            hiddenTreasures.add(t);
        }
        if(number == 6){
            t = dealer.nextTreasure();
            hiddenTreasures.add(t);
        }
    }
    
    public int getLevels(){
        return level;
    }
    
    public Treasure stealTreasure(){
        Treasure t = null;
        if(canISteal()){
            if(enemy.canYouGiveMeATreasure()){
                t = enemy.giveMeATreasure();
                hiddenTreasures.add(t);
                haveStolen();
            }
        }
        return t;
    }
    
    public void setEnemy(Player enemy){
        this.enemy = enemy;
    }
    
    private Treasure giveMeATreasure(){
        return hiddenTreasures.get((int) (Math.random() * (hiddenTreasures.size()-1)));
    }
    
    public boolean canISteal(){
        return canISteal;
    }
    
    private boolean canYouGiveMeATreasure(){
        return hiddenTreasures.size()>0;
    }
    
    private void haveStolen(){
        canISteal = false;
    }
    
    public void discardAllTreasures(){
        ArrayList<Treasure> tr_vis_aux= new ArrayList(visibleTreasures);
        ArrayList<Treasure> tr_hid_aux= new ArrayList(hiddenTreasures);
        
        for(Treasure tr_v :tr_vis_aux)
            discardVisibleTreasure(tr_v);
        
        for(Treasure tr_h :tr_hid_aux)
            discardHiddenTreasure(tr_h);
    }
    
    public String toString(){
        return name + "\n\tNivel = " + level + "\n\tNivel de combate = " + getCombatLevel();
    }
}

File: Ruby/Napakalaki/pkg/pruebaNapakalaki-0.0.1/lib/Player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'Dice.rb'

require_relative 'Dice.rb'
require_relative 'CardDealer.rb'
require_relative 'TreasureKind.rb'
require_relative 'Monster.rb'
require_relative 'CombatResult.rb'
require_relative 'SpecificBadConsequence.rb'
require_relative 'NumericBadConsequence.rb'
require_relative 'DeathBadConsequence.rb'

class Player
  @@MAXLEVEL=10
  
  def initialize(name_p, lvl, dd, steal, ene, hidT, visT, pbad)    
    @name = name_p
    @level = lvl
    @dead = dd
    @canISteal = steal
    @enemy = ene
    @hiddenTreasures = hidT
    @visibleTreasures = visT
    @pendingBadConsequence = pbad
  end
  
  def self.copyPlayer(player)
    new(player.name, player.level, player.dead, player.canISteal, player.enemy, player.hiddenTreasures, player.visibleTreasures, player.pendingBadConsequence)
  end
  
  def bringToLife
    @dead= false
  end
  
  def getCombatLevel
    level_combat= @level
    
    if @visibleTreasures != nil   
      for i in(0..@visibleTreasures.length-1)
        level_combat+= @visibleTreasures[i].bonus
      end
    end
    
    return level_combat
  end
  
  def getOponentLevel(monster)
    monster.combatLevel
  end
  
  def shouldConvert
    dice = Dice.instance
    return dice.nextNumber == 6
  end
  
  def incrementLevels(i)
    @level += i.to_i
    if @level >= @@MAXLEVEL
      @level = 10
    end
  end
  
  def decrementLevels(i)
    @level -= i.to_i
    if @level <= 0
      @level = 1
    end
  end
  
  def setPendingBadConsequence(b)
    @pendingBadConsequence = b
  end
  
  def applyPrize(m)
    nLevels = m.getLevelsGained
    n_treasures = m.getTreasuresGained

    incrementLevels(nLevels)

    if n_treasures>0
      dealer = CardDealer.instance
      for i in 0..n_treasures
        tr_aux = dealer.nextTreasure
        @hiddenTreasures<<(tr_aux)
      end
    end
  end
  
  def applyBadConsequence(m)
    badC = m.bc
    nLevels= badC.levels
        
    decrementLevels(nLevels)

    pendingBad = badC.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
    setPendingBadConsequence(pendingBad)
  end
  
  def canMakeTreasureVisible(t)
    puede_make = true
    n_tr_oh = 0
    @visibleTreasures.each do |tes|
      if tes.type == t.type && t.type != TreasureKind::ONEHAND
        return false
      end
      if tes.type == TreasureKind::ONEHAND
          n_tr_oh += 1

          if t.type == TreasureKind::BOTHHANDS
            return false
          end
      end
      if tes.type == TreasureKind::BOTHHANDS && t.type == TreasureKind::ONEHAND
        return false
      end
    end
    if t.type == TreasureKind::ONEHAND && n_tr_oh >= 2
      return false
    end

    return puede_make
  end
  
  def howManyVisibleTreasures(t)
    n_trs=0
    @visibleTreasures.each do |v_tr|
      if(v_tr==t)
        n_trs+=1
      end
    end
    return n_trs
  end
  
  def dielfNoTreasures
    @dead=true
  end
  
  def giveMeATreasure
    return hiddenTreasures[rand(hiddenTreasures.length)]
  end
  
  def canYouGiveMeATreasure
    return @visibleTreasures.length
  end
  
  def howManyVisibleTreasures(t)
    n_trs = 0
    for i in(0..@visibleTreasures.length-1)
      if visibleTreasures[i].type == t
        n_trs += 1
      end
    end
    return n_trs
  end
  
  def dielfNoTreasures
    if @visibleTreasures.empty? && @hiddenTreasures.empty?
      @dead = true
    end
  end
  
  def getHiddenTreasures
    return @hiddenTreasures
  end
  
  def getVisibleTreasures
    return @visibleTreasures
  end
  
  def combat(m)
    myLevel = getCombatLevel
    monsterLevel = getOponentLevel(m)
    if !canISteal
      dice = Dice.instance
      number = dice.nextNumber

      if number < 3 
        enemyLevel = @enemy.getCombatLevel
        monsterLevel += enemyLevel;
      end
    end
    if myLevel > monsterLevel
      applyPrize(m)
      if level >= @@MAXLEVEL 
        cr = CombatResult::WINGAME
      else
        cr = CombatResult::WIN
      end
    else
      applyBadConsequence(m)
      cr = CombatResult::LOSE
      if shouldConvert
        cr = CombatResult::LOSEANDCONVERT
      end
    end
    
    return cr;
  end
  
  def makeTreasureVisible(t)
    canI = canMakeTreasureVisible(t)
    if canI
      @visibleTreasures.push(t)
      @hiddenTreasures.delete(t)
    end
  end
  
  def discardVisibleTreasure(t)
    @visibleTreasures.delete(t)
    if @pendingBadConsequence != nil && !@pendingBadConsequence.empty? 
      @pendingBadConsequence.substractVisibleTreasure(t)
    end
    dielfNoTreasures()
  end
  
  def discardHiddenTreasure(t)
    @hiddenTreasures.delete(t)
    if @pendingBadConsequence != nil && !@pendingBadConsequence.empty?
      @pendingBadConsequence.substractHiddenTreasure(t)
    end
    dielfNoTreasures()
  end
  
  def validState
    return @hiddenTreasures.length < 5 && (@pendingBadConsequence == nil || @pendingBadConsequence.empty?)
  end
  
  def initTreasures
    dealer = CardDealer.instance
    dice = Dice.instance
    bringToLife
    t = dealer.nextTreasure
    @hiddenTreasures.push(t)
    number = dice.nextNumber

    if(number > 1)
        t = dealer.nextTreasure
        @hiddenTreasures.push(t)
    end
    if(number == 6)
        t = dealer.nextTreasure
        @hiddenTreasures.push(t)
    end
    
    
  end
  
  def getLevels
    return @level;
  end
  
  def stealTreasure
    t = nil;
    if canISteal
      if enemy.canYouGiveMeATreasure
        t = giveMeATreasure
        @hiddenTreasures<<(t)
        haveStolen
      end
    end
    return t
  end
  
  def setEnemy(enemy)
    @enemy = enemy
  end
  
  def giveMeATreasure
    return @hiddenTreasures[rand(@hiddenTreasures.length)]
  end
  
  def canISteal
    return @canISteal
  end
  
  def canYouGiveMeATreasure
    return @visibleTreasures.length > 0
  end
  
  def haveStolen
    @canISteal = false
  end
  
  def discardAllTreasures
    tr_vis_aux = Array.new(@visibleTreasures)
    tr_hid_aux = Array.new(@hiddenTreasures)

    tr_vis_aux.each do |v_tr|
      discardVisibleTreasure(v_tr)
    end

    tr_hid_aux.each do |h_tr|
      discardHiddenTreasure(h_tr)
    end
  end
  
  def to_s
    "#{@name}
    \tNivel = #{@level}
    \tNivel de combate = " + getCombatLevel.to_s
  end
  
  def self.MAXLEVEL
    return @@MAXLEVEL
  end
  
  attr_accessor :canISteal, :level, :name, :dead, :enemy, :pendingBadConsequence, :hiddenTreasures, :visibleTreasures
  private :bringToLife, :getCombatLevel, :incrementLevels, :decrementLevels,
    :setPendingBadConsequence, :applyPrize, :applyBadConsequence, :canMakeTreasureVisible,
    :howManyVisibleTreasures, :dielfNoTreasures, :giveMeATreasure, :canYouGiveMeATreasure,
    :haveStolen
end